timemachine/cpp/src/context.cu

```diff
@@ -74,65 +61,65 @@ Context::multiple_steps(const std::vector<double> &lambda_schedule, int store_du
     int box_buffer_size = x_buffer_size * 3 * 3;
 
     std::vector<double> h_x_buffer(x_buffer_size * N_ * 3);
+
     cudaStream_t stream = static_cast<cudaStream_t>(0);
-    try {
-        gpuErrchk(cudaMalloc(&d_box_buffer, box_buffer_size * sizeof(*d_box_buffer)));
-        // indicator so we can set it to a default arg.
-        gpuErrchk(cudaMalloc(&d_du_dl_buffer, du_dl_buffer_size * sizeof(*d_du_dl_buffer)));
-        gpuErrchk(cudaMemset(d_du_dl_buffer, 0, du_dl_buffer_size * sizeof(*d_du_dl_buffer)));
-        intg_->initialize(bps_, lambda_schedule[0], d_x_t_, d_v_t_, d_box_t_, stream);
-        for (int i = 1; i <= lambda_schedule.size(); i++) {
-            // decide if we need to store the du_dl for this step
-            unsigned long long *du_dl_ptr = nullptr;
-            if (i % store_du_dl_interval == 0) {
-                // pemdas but just to make it clear we're doing pointer arithmetic
-                du_dl_ptr = d_du_dl_buffer + ((i / store_du_dl_interval) - 1);
-            }
 
-            double lambda = lambda_schedule[i - 1];
-            this->_step(bps_, lambda, du_dl_ptr, stream);
-
-            if (i % store_x_interval == 0) {
-                gpuErrchk(cudaMemcpy(
-                    &h_x_buffer[0] + ((i / store_x_interval) - 1) * N_ * 3,
-                    d_x_t_,
-                    N_ * 3 * sizeof(double),
-                    cudaMemcpyDeviceToHost));
-                gpuErrchk(cudaMemcpy(
-                    &d_box_buffer[0] + ((i / store_x_interval) - 1) * 3 * 3,
-                    d_box_t_,
-                    3 * 3 * sizeof(*d_box_buffer),
-                    cudaMemcpyDeviceToDevice));
-            }
+    std::unique_ptr<DeviceBuffer<double>> d_box_buffer(nullptr);
+    if (box_buffer_size > 0) {
+        d_box_buffer.reset(new DeviceBuffer<double>(box_buffer_size));
+    }
+    std::unique_ptr<DeviceBuffer<unsigned long long>> d_du_dl_buffer(nullptr);
+    if (du_dl_buffer_size > 0) {
+        d_du_dl_buffer.reset(new DeviceBuffer<unsigned long long>(du_dl_buffer_size));
+        gpuErrchk(cudaMemsetAsync(d_du_dl_buffer->data, 0, d_du_dl_buffer->size, stream));
+    }
```

Comment:
Just noticing that we could avoid `unique_ptr` and dynamic allocation here except that `DeviceBuffer` currently has a check that fails if the size is zero. Would it be worth relaxing that restriction? (conflicts with e.g. the behavior of `malloc` AFAIK)

Reply:
I would be for removing the check as it is consistent with cudaMalloc's behavior (nullptr if size is 0). Last time this came up though we decided against it, though I can't find a reference to that conversation.

Easily done in this PR and would simplify a fair amount of code.

Reply:
To be clear, I wouldn't argue for using a literal `DeviceBuffer(0)` in cases where we don't want to allocate (I think the discussion starting [here](https://github.com/proteneer/timemachine/pull/726#discussion_r853317229) was more about that case)

Reply:
1) I agree that wrapping a `DeviceBuffer` in a `unique_ptr` is unwieldy.
2) Do we actually have any practical situations where `box_buffer_size`, and `du_dl_buffer_size` is zero? If not, I'd just turn assert that those two quantities are `> 0` and simplify.